factom/client.py

```diff
@@ -301,31 +301,58 @@ def transaction(self, tx_hash: Union[bytes, str]):
-        """
-        entries = []
+    def read_chain(
+        self,
+        chain_id: Union[bytes, str],
+        from_height: int = 0,
+        include_entry_context: bool = False,
+        encode_as_hex: bool = False,
+    ):
+        """A generator that yields all entries of a chain in order, optionally starting from a given block height."""
+        # Walk the entry block chain backwards to build up a stack of entry blocks to fetch
+        entry_blocks = []
         keymr = self.chain_head(chain_id)["chainhead"]
         while keymr != NULL_BLOCK:
             block = self.entry_block(keymr)
-            for entry_pointer in reversed(block["entrylist"]):
+            if block["header"]["dbheight"] < from_height:
+                break
+            entry_blocks.append(block)
+            keymr = block["header"]["prevkeymr"]
+
+        # Continuously pop off the stack and yield each entry one by one (in the order that they appear in the block)
+        while len(entry_blocks) > 0:
+            entry_block = entry_blocks.pop()
+            for entry_pointer in entry_block["entrylist"]:
                 entry = self.entry(entry_pointer["entryhash"], encode_as_hex=encode_as_hex)
                 if include_entry_context:
                     entry["entryhash"] = entry_pointer["entryhash"]
                     entry["timestamp"] = entry_pointer["timestamp"]
-                    entry["dbheight"] = block["header"]["dbheight"]
-                entries.append(entry)
-            keymr = block["header"]["prevkeymr"]
-        return entries
+                    entry["dbheight"] = entry_block["header"]["dbheight"]
+                yield entry
+
+    def entries_at_height(
+        self, chain_id: Union[bytes, str], height: int, include_entry_context: bool = False, encode_as_hex: bool = False
+    ):
+        """A generator that yields all entries in a chain that occurred at the given height."""
+        # Look for the chain id in the directory block entries
+        target_chain_id = utils.hex_from_bytes_or_string(chain_id)
+        directory_block = self.directory_block_by_height(height)["dblock"]
+        for entry_block_pointer in directory_block["dbentries"]:
+            if entry_block_pointer["chainid"] == target_chain_id:
+                entry_block_keymr = entry_block_pointer["keymr"]
+                break
+        else:
+            return []  # Early return, chain didn't have entries in this block
+
+        # Entry block found, yield all entries within the block
+        entry_block = self.entry_block(entry_block_keymr)
+        for entry_pointer in entry_block["entrylist"]:
+            entry = self.entry(entry_pointer["entryhash"], encode_as_hex=encode_as_hex)
+            if include_entry_context:
+                entry["entryhash"] = entry_pointer["entryhash"]
+                entry["timestamp"] = entry_pointer["timestamp"]
+                entry["dbheight"] = entry_block["header"]["dbheight"]
+            yield entry
```

Comment:
I think it might be a good idea to move:

```python
for entry_pointer in entry_block["entrylist"]:
    entry = self.entry(entry_pointer["entryhash"], encode_as_hex=encode_as_hex)
    if include_entry_context:
        entry["entryhash"] = entry_pointer["entryhash"]
        entry["timestamp"] = entry_pointer["timestamp"]
        entry["dbheight"] = entry_block["header"]["dbheight"]
    yield entry
```
into a helper method `gen_entries_in_block(entry_block, include_entry_context=False)` and then we could get rid of the code duplication in this method and `read_chain`?

Reply:
Yeah that's a really good idea, useful function to expose to users too I think